server/routes/experiments/biomed_nl/traversal.py

```diff
@@ -0,0 +1,536 @@
+  def initialize_models(self,
+                        gemini_api_key,
+                        gemini_model_str='gemini-2.0-flash-001'):
+    self.embeddings_model = SentenceTransformer(
+        'all-MiniLM-L6-v2')  # Or another lightweight model
+    self.gemini = genai.Client(
+        api_key=gemini_api_key,
+        http_options=genai.types.HttpOptions(api_version='v1alpha'))
+    self.gemini_model_str = gemini_model_str
+
+  def traverse_n_hops(self, start_dcids, n):
+    '''Traverses the graph for a specified number of hops, updating the path store.
+
+    Args:
+        start_dcids: A list of starting DCIDs.
+        n: The number of hops to traverse.
+    '''
+    dcids = start_dcids
+    for _ in range(n):
+      if not dcids:
+        return
+
+      out_triples = get_next_hop_triples(dcids, out=True)
+      in_triples = get_next_hop_triples(dcids, out=False)
+      triples = {
+          dcid: {
+              **out_triples.get(dcid, {}),
+              **in_triples.get(dcid, {})
+          } for dcid in set(out_triples.keys()) | set(in_triples.keys())
+      }
+
+      self.path_store.merge_triples_into_path_store(triples)
+      self.path_store.sample_next_hops(MIN_SAMPLE_SIZE)
+      dcids = self.path_store.get_next_dcids()
+
+  def filter_paths_with_embeddings_model(self, pct=0.1):
+    '''Filters paths based on cosine similarity between property descriptions 
+    and the query.
+
+    Args:
+        pct: The percentage of top properties to keep (e.g., 0.1 for 10%).
+
+    Returns:
+        A list of the top properties that were selected for filtering.
+    '''
+
+    property_descriptions = self.path_store.get_property_descriptions()
+
+    # TODO: add a description to description in DC KG.
+    property_descriptions['description'] = (
+        'The description describes the entity by its characteristics or '
+        'attributes, providing information about that entity to help to define '
+        'and distinguish it.')
+
+    prop_embeds = {
+        prop:
+            self.embeddings_model.encode(property_descriptions[prop],
+                                         convert_to_tensor=True)
+        for prop in property_descriptions
+    }
+    query_embed = self.embeddings_model.encode(self.query,
+                                               convert_to_tensor=True)
+
+    properties_to_similarity_score = {
+        prop: util.cos_sim(query_embed, prop_embed).item()
+        for prop, prop_embed in prop_embeds.items()
+    }
+    num_top_properties = math.ceil(len(properties_to_similarity_score) * pct)
+    top_props = [
+        property[0]
+        for property in sorted(properties_to_similarity_score.items(),
+                               key=lambda item: item[1],
+                               reverse=True)[:num_top_properties]
+    ]
+
+    self.path_store.filter_by_properties(top_props)
+
+    return top_props
+
+  def select_paths(self):
+    '''Selects paths to purpose using the Gemini model based on a prompt.
```

Comment:
what does paths to purpose mean?

Reply:
typo -- I meant paths to _pursue_. Thanks for catching!!